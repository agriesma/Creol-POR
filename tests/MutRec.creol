/*
 * An example displaying a common dead-lock situation
 *
 * Marcel Kyas, 2007.
 *
 * The contents of this file is hereby placed into the public domain.
 *
 * This file is part of creoltools.
 */
class MutRec
implements MutRec
begin
  op init == skip
  op run == skip
  op doit(in other: MutRec) == other.doit(this;)
end

class Test
begin
  var m1: MutRec
  var m2: MutRec

  op init ==  m1 := new MutRec(); m2 := new MutRec()
  op run == !m1.doit(m2)
end
