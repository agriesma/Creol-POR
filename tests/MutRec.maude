load creol-interpreter
mod PROGRAM is
pr CREOL-INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "MutRec" : Cl | Inh: noInh, Par: noVid, Att: noSubst, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "doit" : Mtdname | Param: "other", Latt: "other" |-> null, Code: "sync.0" ! "other" . "doit" ( "this" ) ; ( "sync.0" ? ( noVid ) )  ; return ( emp ) >, Ocnt: 0 >

< "Test" : Cl | Inh: noInh, Par: noVid, Att: "m1" |-> null , "m2" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: "m1" ::= new "MutRec" ( emp ) ; "m2" ::= new "MutRec" ( emp ) ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: ".anon" ! "m1" . "doit" ( "m2" ) ; free( ".anon" ) ; return ( emp ) >, Ocnt: 0 >

.
endm
