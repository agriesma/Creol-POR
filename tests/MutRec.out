==========================================
reduce in PROGRAM : init .
result Configuration: 
< "MutRec" : Cl | Inh: noInh, Par: noVid, Att: noSubst, Mtds:
  < "doit" : Mtdname | Param: "other", Latt: "other" |-> null, Code: "sync.0" !
    "other" . "doit"("this") ; "sync.0" ?(noVid) ; return(emp) > *
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >,
  Ocnt: 0 >

< "Test" : Cl | Inh: noInh, Par: noVid, Att: "m1" |-> null, "m2" |-> null,
    Mtds:
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: "m1" ::= new
    "MutRec"(emp) ; "m2" ::= new "MutRec"(emp) ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: ".anon" ! "m1" .
    "doit"("m2") ; free(".anon") ; return(emp) >,
  Ocnt: 0 >
Bye.
